lib/metadata/bucketfile/backend.js

```diff
@@ -0,0 +1,451 @@
+        const requestParams = {};
+        let Ext;
+        let extParams = {};
+        // multipart upload listing
+        if (params.listingType === 'multipartuploads') {
+            Ext = arsenal.listMPU.ListMultipartUploads;
+            this.addExtensionParam('queryPrefixLength', params, extParams);
+            this.addExtensionParam('splitter', params, extParams);
+            if (params.keyMarker) {
+                requestParams.gt = `overview${params.splitter}` +
+                    `${params.keyMarker}${params.splitter}`;
+                if (params.uploadIdMarker) {
+                    requestParams.gt += `${params.uploadIdMarker}`;
+                }
+                // advance so that lower bound does not include the supplied
+                // markers
+                requestParams.gt = this._setCharAt(requestParams.gt);
+            }
+        } else {
+            Ext = arsenal.delimiter.Delimiter;
+            if (params.marker) {
+                requestParams.gt = params.marker;
+                this.addExtensionParam('gt', requestParams, extParams);
+            }
+        }
+        this.addExtensionParam('delimiter', params, extParams);
+        this.addExtensionParam('maxKeys', params, extParams);
+        if (params.prefix) {
+            requestParams.start = params.prefix;
+            requestParams.lt = this._setCharAt(params.prefix);
+            this.addExtensionParam('start', requestParams, extParams);
+        }
+        const extension = new Ext(extParams, log);
+        this.loadDBIfExists(bucketName, log, (err, db) => {
+            if (err) {
+                return cb(err);
+            }
+            let cbDone = false;
+            const stream = db.createReadStream(requestParams);
+            stream
+                .on('data', e => {
+                    if (extension.filter(e) === false) {
+                        stream.emit('end');
+                        stream.destroy();
+                    }
+                })
+                .on('error', err => {
+                    if (!cbDone) {
+                        this.unRef();
+                        cbDone = true;
+                        log.error('error listing objects',
+                                  { error: err });
+                        cb(errors.InternalError);
+                    }
+                })
+                .on('end', () => {
+                    if (!cbDone) {
+                        this.unRef();
+                        cbDone = true;
+                        const data = extension.result();
+                        cb(null, data);
+                    }
+                });
+            return undefined;
+        });
+    }
+
+    listObject(bucketName, params, log, cb) {
+        return this.internalListObject(bucketName, params, log, cb);
+    }
+
+    listMultipartUploads(bucketName, params, log, cb) {
+        return this.internalListObject(bucketName, params, log, cb);
```

Comment:
The internalListObject() will use params.listingtype to differenciate different behavior. The fact that we use a common function to deal with many cases is specific to our implementation, but could be different in some other implementations of the backend class.